server/manager/src/main/java/org/apache/accumulo/manager/BalanceManager.java

```diff
@@ -0,0 +1,331 @@
+import org.apache.accumulo.core.metrics.MetricsProducer;
+import org.apache.accumulo.core.spi.balancer.BalancerEnvironment;
+import org.apache.accumulo.core.spi.balancer.TableLoadBalancer;
+import org.apache.accumulo.core.spi.balancer.TabletBalancer;
+import org.apache.accumulo.core.spi.balancer.data.TServerStatus;
+import org.apache.accumulo.core.spi.balancer.data.TabletMigration;
+import org.apache.accumulo.core.spi.balancer.data.TabletServerId;
+import org.apache.accumulo.manager.metrics.BalancerMetrics;
+import org.apache.accumulo.server.ServerContext;
+import org.apache.accumulo.server.manager.balancer.BalancerEnvironmentImpl;
+import org.apache.accumulo.server.manager.state.UnassignedTablet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+public class BalanceManager {
+
+  private static final Logger log = LoggerFactory.getLogger(BalanceManager.class);
+
+  private final Manager manager;
+  protected volatile TabletBalancer tabletBalancer;
+  private final BalancerEnvironment balancerEnvironment;
+  private final BalancerMetrics balancerMetrics = new BalancerMetrics();
+  private final Object balancedNotifier = new Object();
+
+  BalanceManager(Manager manager) {
+    this.manager = manager;
+    this.balancerEnvironment = new BalancerEnvironmentImpl(manager.getContext());
+    initializeBalancer();
+  }
+
+  private void initializeBalancer() {
+    var localTabletBalancer =
+        Property.createInstanceFromPropertyName(getContext().getConfiguration(),
+            Property.MANAGER_TABLET_BALANCER, TabletBalancer.class, new TableLoadBalancer());
+    localTabletBalancer.init(balancerEnvironment);
+    tabletBalancer = localTabletBalancer;
+  }
+
+  void propertyChanged(String property) {
+    String resolved = DeprecatedPropertyUtil.getReplacementName(property, (log, replacement) -> {});
+    if (resolved.equals(Property.MANAGER_TABLET_BALANCER.getKey())) {
+      initializeBalancer();
+      log.info("tablet balancer changed to {}", tabletBalancer.getClass().getName());
+    }
+  }
+
+  private ServerContext getContext() {
+    return manager.getContext();
+  }
+
+  MetricsProducer getMetrics() {
+    return balancerMetrics;
+  }
+
+  /**
+   * balanceTablets() balances tables by DataLevel. Return the current set of migrations partitioned
+   * by DataLevel
+   */
+  private Map<Ample.DataLevel,Set<KeyExtent>> partitionMigrations() {
+    final Map<Ample.DataLevel,Set<KeyExtent>> partitionedMigrations =
+        new EnumMap<>(Ample.DataLevel.class);
+    for (Ample.DataLevel dl : Ample.DataLevel.values()) {
+      Set<KeyExtent> extents = new HashSet<>();
+      // prev row needed for the extent
+      try (
+          var tabletsMetadata = getContext()
+              .getAmple().readTablets().forLevel(dl).fetch(TabletMetadata.ColumnType.PREV_ROW,
+                  TabletMetadata.ColumnType.LOCATION, TabletMetadata.ColumnType.MIGRATION)
```

Comment:
Added the LOCATION column to this code because I noticed an error in the logs where shouldCleanupMigration() was throw an exception because the column was not fetched.

Reply:
When I merged main and resolved conflicts in 35f482b1bde21c3d66c3f3ebda769e2cb9958c22 I copied the latest versions of any moved code in the manager over, so got those changes.